crates/red_knot/src/module.rs

```diff
@@ -0,0 +1,708 @@
+                // ```
+                // src/foo.py
+                // src/foo/__init__.py
+                // ```
+                // The module name of `src/foo.py` is `foo`, but the module loaded by Python is `src/foo/__init__.py`.
+                // That means we need to ignore `src/foo.py` even though it resolves to the same module name.
+                return None;
+            }
+
+            // Path has been inserted by `resolved`
+            Some(module_id)
+        } else {
+            // This path is for a module with the same name but in a module search path with a lower priority.
+            // Ignore it.
+            None
+        }
+    }
+
+    pub fn add_module(&mut self, path: &Path) {
+        self.resolve_path(path);
+    }
+
+    pub fn remove_module(&mut self, id: ModuleId) {
+        match self.modules.entry(id) {
+            Entry::Occupied(entry) => {
+                // Keep the entry to ensure it locks any writes to the same entry in `resolve`.
+
+                self.by_name.remove(&entry.get().name);
+                // This isn't fast but removing seems an uncommon operation where I don't think it's worth
+                // to keep a reverse lookup just for it (requires more allocations, involves writing a lot of data).
+                self.by_path.retain(|_, current_id| *current_id != id);
+
+                entry.remove();
+            }
+            Entry::Vacant(_) => {
+                panic!("Expected entry to exist")
+            }
+        }
+    }
+}
+
+impl std::fmt::Debug for ModuleResolver {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("ModuleResolver")
+            .field("search_paths", &self.search_paths)
+            .field("modules", &self.by_name)
+            .finish()
+    }
+}
+
+fn resolve_name(
+    name: &ModuleName,
+    search_paths: &[ModuleSearchPath],
+) -> Option<(ModuleSearchPath, PathBuf)> {
+    'search_path: for search_path in search_paths {
+        let mut package_path = search_path.path().to_path_buf();
+
+        let mut components = name.components();
+        let module_name = components.next_back()?;
+
+        // For `foo.bar.baz`, test that `foo` and `baz` both contain a `__init__.py`.
+        for folder in components {
+            package_path.push(folder);
+
+            // A folder on the path to the module. Test that it contains a `__init__.py`.
+            if !package_path.join("__init__.py").is_file()
+                && !package_path.join("__init__.pyi").is_file()
+            {
+                // Try the next search path
+                continue 'search_path;
+            }
+        }
+
+        package_path.push(module_name);
+
+        // Must be a `__init__.pyi` or `__init__.py` or it isn't a package.
+        if package_path.is_dir() {
+            package_path.push("__init__");
+        }
+
+        let stub = package_path.with_extension("pyi");
+
+        if stub.is_file() {
```

Comment:
I'm not sure I understand. What other use cases do we have for getting the list of candidate paths? 

I do agree that we should use the VFS here. But I intentionally didn't build this out yet and used the real file system as a pretty good mock ;)

>  (And ideally I think "check if file exists" and "read file contents" should be a single atomic operation to avoid race conditions; or at least they should always happen in the same method, one right after the other, so we can handle a read error the same as "doesn't exist" )

See my comment above.

Reply:
Ok, I buy that we will need to be able to resolve modules without necessarily reading their contents.

It sounds like you are already planning that in future we should ensure that all filesystem access happens through VFS; in that case, we are on the same page. E.g. I want to be able to write tests for the logic of `ModuleResolver` using a mock VFS that don't require actual filesystem access.